x/concentrated-liquidity/fees.go

```diff
@@ -255,3 +255,45 @@ func calculateFeeGrowth(targetTick int64, feeGrowthOutside sdk.DecCoins, current
 func formatPositionAccumulatorKey(poolId uint64, owner sdk.AccAddress, lowerTick, upperTick int64) string {
 	return strings.Join([]string{strconv.FormatUint(poolId, uintBase), owner.String(), strconv.FormatInt(lowerTick, uintBase), strconv.FormatInt(upperTick, uintBase)}, keySeparator)
 }
+
+// computeFeeChargePerStep returns the total fee charge per swap step given the parameters.
+// - currentSqrtPrice the sqrt price at which the swap step begins.
+// - nextSqrtPrice the sqrt price at which the swap step should end.
+// - sqrtPriceLimit the sqrt price corresponding to the sqrt of the price representing price impact protection.
+// - amountIn the amount of token in to be consumed during the swap step
+// - amountSpecifiedRemaining is the total remaining amount of token in that needs to be consumed to complete the swap.
+// - swapFee the swap fee to be charged.
+//
+// If swap fee is negative, it panics.
+// If swap fee is 0, returns 0. Otherwise, computes and returns the fee charge per step.
+// TODO: test this function.
+func computeFeeChargePerStep(currentSqrtPrice, nextSqrtPrice, sqrtPriceLimit, amountIn, amountSpecifiedRemaining, swapFee sdk.Dec) sdk.Dec {
+	feeChargeTotal := sdk.ZeroDec()
+
+	if swapFee.IsNegative() {
+		// This should never happen but is added as a defense-in-depth measure.
+		panic(fmt.Errorf("swap fee must be non-negative, was (%s)", swapFee))
+	}
+
+	if swapFee.IsZero() {
+		return feeChargeTotal
+	}
+
+	// 1. The current tick does not have enough liqudity to fulfill the swap.
+	didReachNextSqrtPrice := nextSqrtPrice.Equal(currentSqrtPrice)
+	// 2. The next sqrt price was not reached due to price impact protection.
+	isPriceImpactProtection := currentSqrtPrice.Equal(sqrtPriceLimit)
```

Comment:
Will get back to this

Reply:
Mmm, I feel like the way it currently is makes sense. The currentSqrtPrice will never exceed the sqrtPriceLimit AFAIK and when we hit it they are identical, so seems like this would be a logical boolean trigger for this.

Reply:
Also, here's the breakdown of this logic:
https://github.com/osmosis-labs/osmosis/pull/4097#discussion_r1087470350

Reply:
@AlpinYukseloglu it should be equal here since we check for GT  prior in the execution flow and the swap all the way until `sqrtPriceLimit`:

https://github.com/osmosis-labs/osmosis/blob/4d9b21299a8e2942919a17d3a1f25335c36cd964/x/concentrated-liquidity/internal/swapstrategy/one_for_zero.go#L34-L36

If price impact was triggered, `currentSqrtPrice` must equal `sqrtPriceLimit` here